qualtran/bloqs/chemistry/thc.ipynb

```diff
@@ -0,0 +1,291 @@
+  {
+   "cell_type": "markdown",
+   "id": "5b253057",
+   "metadata": {
+    "cq.autogen": "_make_prepare.md"
+   },
+   "source": [
+    "## `PrepareTHC`\n",
+    "State Preparation for THC Hamilontian.\n",
+    "\n",
+    "Prepares the state\n",
+    "\n",
+    "$$\n",
+    "    \\frac{1}{\\sqrt{\\lambda}}|+\\rangle|+\\rangle\\left[\n",
+    "        \\sum_\\ell^{N/2} \\sqrt{t_\\ell}|\\ell\\rangle|M+1\\rangle\n",
+    "        + \\frac{1}{\\sqrt{2}} \\sum_{\\mu\\le\\nu}^M \\sqrt{\\zeta_{\\mu\\nu}} |\\mu\\rangle|\\nu\\rangle\n",
+    "    \\right].\n",
+    "$$\n",
+    "\n",
+    "Note we use UniformSuperpositionTHC as a subroutine as part of this bloq in\n",
+    "contrast to the reference which keeps them separate.\n",
+    "\n",
+    "#### Parameters\n",
+    " - `num_mu`: THC auxiliary index dimension $M$\n",
+    " - `num_spin_orb`: number of spin orbitals $N$\n",
+    " - `alt_mu`: Alternate values for mu indices.\n",
+    " - `alt_nu`: Alternate values for nu indices.\n",
+    " - `alt_theta`: Alternate values for theta indices.\n",
+    " - `theta`: Signs of lcu coefficients.\n",
+    " - `keep`: keep values.\n",
+    " - `keep_bitsize`: number of bits for keep register for coherent alias sampling. \n",
+    "\n",
+    "Registers:\n",
+    " - mu: $\\mu$ register.\n",
+    " - nu: $\\nu$ register.\n",
+    " - theta: sign register.\n",
+    " - succ: success flag qubit from uniform state preparation\n",
+    " - eq_nu_mp1: flag for if $nu = M+1$\n",
+    " - plus_a / plus_b: plus state for controlled swaps on spins.\n",
+    "\n",
+    "#### References\n",
+    "[Even more efficient quantum computations of chemistry through tensor hypercontraction](https://arxiv.org/pdf/2011.03494.pdf) Fig. 2 and Fig. 3.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "c4886f09",
+   "metadata": {
+    "cq.autogen": "_make_prepare.py"
+   },
+   "outputs": [],
+   "source": [
+    "from qualtran.bloqs.chemistry.thc import PrepareTHC\n",
+    "\n",
+    "num_spat = 20\n",
+    "num_mu = 20\n",
+    "t_l = np.random.randint(0, 10, size=num_spat)\n",
+    "zeta = np.random.randint(0, 10, size=(num_mu, num_mu))\n",
+    "zeta = 0.5 * (zeta + zeta.T)\n",
+    "eps = 1e-3\n",
+    "bloq = PrepareTHC.from_hamiltonian_coeffs(t_l, zeta, probability_epsilon=eps)\n",
+    "show_bloq(bloq.decompose_bloq())"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "ec8f8b91",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from qualtran.resource_counting import get_bloq_counts_graph\n",
+    "from qualtran.drawing import GraphvizCounts\n",
+    "from cirq_ft.algos.multi_control_multi_target_pauli import MultiControlPauli\n",
+    "\n",
+    "graph, sigma = get_bloq_counts_graph(bloq, generalizer=generalize)\n",
+    "GraphvizCounts(graph).get_svg()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "85818cf5",
+   "metadata": {},
+   "source": [
+    "### ```Paper Comparison```\n",
```

Comment:
shouldn't be a code block

Reply:
It's markdown?

Reply:
you've got your heading in backticks again

Reply:
Ah